cmd/bedEnrichments/bedEnrichments.go

```diff
@@ -0,0 +1,160 @@
+package main
+
+import (
+	"flag"
+	"fmt"
+	"github.com/vertgenlab/gonomics/bed"
+	"github.com/vertgenlab/gonomics/exception"
+	"github.com/vertgenlab/gonomics/fileio"
+	"github.com/vertgenlab/gonomics/interval"
+	"log"
+)
+
+func bedEnrichments(method string, inFile string, secondFile string, noGapFile string, outFile string, verbose int, trimToRefGenome bool) {
+	var err error
+
+	if method != "exact" && method != "normalApproximate" && method != "upperBound" && method != "lowerBound" {
+		log.Fatalf("Error: unknown method. Found: %s.", method)
+	}
+
+	if verbose > 0 {
+		log.Println("Reading bed files.")
+	}
+
+	elementsOne := bed.ReadLite(inFile)
+	elementsTwo := bed.ReadLite(secondFile)
+	noGapRegions := bed.ReadLite(noGapFile)
+
+	if trimToRefGenome {
+		var overlap1, overlap2 []interval.Interval
+		var trimmedE1 []bed.Bed = make([]bed.Bed, 0)
+		var trimmedE2 []bed.Bed = make([]bed.Bed, 0)
+
+		var e1Intervals []interval.Interval
+		for i := range elementsOne {
+			e1Intervals = append(e1Intervals, elementsOne[i])
+		}
+		tree1 := interval.BuildTree(e1Intervals)
+		var e2Intervals []interval.Interval
+		for i := range elementsTwo {
+			e2Intervals = append(e2Intervals, elementsTwo[i])
+		}
+		tree2 := interval.BuildTree(e2Intervals)
+
+		for i := range noGapRegions {
+			overlap1 = interval.Query(tree1, noGapRegions[i], "within")
+			if overlap1 != nil {
+				for j := range overlap1 {
+					trimmedE1 = append(trimmedE1, overlap1[j].(bed.Bed))
+				}
+			}
+			overlap2 = interval.Query(tree2, noGapRegions[i], "within")
+			if overlap2 != nil {
+				for j := range overlap2 {
+					trimmedE2 = append(trimmedE2, overlap2[j].(bed.Bed))
+				}
+			}
+		}
+		elementsOne = trimmedE1
+		elementsTwo = trimmedE2
+	}
+
+	if verbose > 0 {
+		log.Println("Sorting bed files.")
+	}
+
+	bed.SortByCoord(elementsOne)
+	bed.SortByCoord(elementsTwo)
+	bed.SortByCoord(noGapRegions)
+
+	if verbose > 0 {
+		log.Println("Running preflight checks.")
+	}
+
+	//preflight checks: check for error in user input. Beds should not be self-overlapping.
+	if bed.IsSelfOverlapping(elementsOne) {
+		log.Fatalf("Elements in bedEnrichments must not be self-overlapping. Self-overlap found in %s.", inFile)
+	}
+	if bed.IsSelfOverlapping(elementsTwo) {
+		log.Fatalf("Elements in bedEnrichments must not be self-overlapping. Self-overlap found in %s.", secondFile)
+	}
+	if bed.IsSelfOverlapping(noGapRegions) {
+		log.Fatalf("Elements in bedEnrichments must not be self-overlapping. Self-overlap found in %s.", noGapFile)
+	}
+
+	var summarySlice []float64
+	overlapCount := bed.OverlapCount(elementsOne, elementsTwo)
+
+	if verbose > 0 {
+		log.Println("Calculating enrichment probabilities.")
+	}
+
+	switch method {
+	case "exact":
+		probs := bed.ElementOverlapProbabilities(elementsOne, elementsTwo, noGapRegions)
```

Comment:
I think we need to double-check these function calls.  I am worried that the above call to OverlapCount gives us a max of len(elementsOne) as the number of overlaps, while the call to ElementOverlapProbabilities is designed such that it is counting the number of elementsTwo that overlap anything in elementsOne.  Is that correct?  If so, we could just swap the args to OverlapCount